File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.go
type NumMatrix struct {
  sums [][]int  
}


func Constructor(matrix [][]int) NumMatrix {
  m, n := len(matrix), len(matrix[0])
  sums := make([][]int, m+1)
  sums[0] = make([]int, n+1)
  for i, r := range matrix {
    row := make([]int, n+1)
    sums[i+1] = row
    for j, v := range r {
      row[j+1] = v + sums[i][j+1] + row[j] - sums[i][j]
    }
  }
  return NumMatrix{sums}  
}

func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
  return this.sums[row2+1][col2+1] - this.sums[row2+1][col1] - this.sums[row1][col2+1] + this.sums[row1][col1]
}


/**
 * Your NumMatrix object will be instantiated and called as such:
 * obj := Constructor(matrix);
 * param_1 := obj.SumRegion(row1,col1,row2,col2);
 */

File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.js
/**
 * @param {number[][]} matrix
 */
var NumMatrix = function(matrix) {
  this.matrix = matrix
  for (j = 1; j < matrix[0].length; j++) {    
    this.matrix[0][j] = this.matrix[0][j] + this.matrix[0][j-1] 
  }

  for (i = 1; i < matrix.length; i++) {
    this.matrix[i][0] = this.matrix[i][0] + this.matrix[i-1][0]
    
    
    for (j = 1; j < matrix[0].length; j++) {
    
    this.matrix[i][j] = this.matrix[i - 1][j] 
      + this.matrix[i][j-1] 
      - this.matrix[i-1][j-1] 
      + this.matrix[i][j]
      
    }
  }
  
    
    
    
  /*[[1,1,1,1,1],
     [2,2,2,2,2],
     [3,123,123,12,3],
     [4,123,12312,12,1]] */

};

/** 
 * @param {number} row1 
 * @param {number} col1 
 * @param {number} row2 
 * @param {number} col2
 * @return {number}
 */
NumMatrix.prototype.sumRegion = function(row1, col1, row2, col2) {
  if (row1 === 0 && col1 === 0) { 
    
    return(this.matrix[row2][col2])
    
  }
  if (row1 === 0) {
    return(this.matrix[row2][col2] - this.matrix[row2][col1-1])
    
  }
  if (col1 === 0) {
    
     return(this.matrix[row2][col2] - this.matrix[row1-1][col2])
  }
  return (this.matrix[row2][col2] - this.matrix[row1-1][col2] - this.matrix[row2][col1-1] + this.matrix[row1-1][col1-1])
};

/** 
 * Your NumMatrix object will be instantiated and called as such:
 * var obj = new NumMatrix(matrix)
 * var param_1 = obj.sumRegion(row1,col1,row2,col2)
 */